File: src/rgbd_throttle.cpp
/*******************************************************************************

Software License Agreement (BSD 3-Clause)

Copyright (c) 2016, Antonio Coratelli
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

1. Redistributions of source code must retain the above copyright notice,
   this list of conditions and the following disclaimer.

2. Redistributions in binary form must reproduce the above copyright notice,
   this list of conditions and the following disclaimer in the documentation
   and/or other materials provided with the distribution.

3. Neither the name of the copyright holder nor the names of its contributors
   may be used to endorse or promote products derived from this software
   without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
POSSIBILITY OF SUCH DAMAGE.

*******************************************************************************/

#include "ros/ros.h"
#include "sensor_msgs/CameraInfo.h"
#include "sensor_msgs/Image.h"

// General Defines
#define NAME "rgbd_throttle"

// Default Param Defines
#define RATE 5.0

// Input Defines
#define RGB_INFO_IN   "rgb/info_in"
#define RGB_RECT_IN   "rgb/rect_in"
#define DEPTH_INFO_IN "depth/info_in"
#define DEPTH_RECT_IN "depth/rect_in"
#define BUFFER_IN 1

// Output Defines
#define RGB_INFO_OUT   "rgb/info_out"
#define RGB_RECT_OUT   "rgb/rect_out"
#define DEPTH_INFO_OUT "depth/info_out"
#define DEPTH_RECT_OUT "depth/rect_out"
#define BUFFER_OUT 1

// Global Variables
ros::Publisher pub_rgb_info;
ros::Publisher pub_rgb_rect;
ros::Publisher pub_depth_info;
ros::Publisher pub_depth_rect;

double rate;
double secs;
double last_sent;

sensor_msgs::CameraInfo rgb_info;
sensor_msgs::Image      rgb_rect;
sensor_msgs::CameraInfo depth_info;
sensor_msgs::Image      depth_rect;

// Callbacks

void callback_rgb_info(const sensor_msgs::CameraInfo& data)
{
    rgb_info = data;
}

void callback_rgb_rect(const sensor_msgs::Image& data)
{
    rgb_rect = data;
}

void callback_depth_info(const sensor_msgs::CameraInfo& data)
{
    depth_info = data;
}

void callback_depth_rect(const sensor_msgs::Image& data)
{
    depth_rect = data;

    if (data.header.stamp.toSec() >= last_sent + secs)
    {
        pub_rgb_info.publish(rgb_info);
        pub_rgb_rect.publish(rgb_rect);
        pub_depth_info.publish(depth_info);
        pub_depth_rect.publish(depth_rect);
        last_sent = data.header.stamp.toSec();
    }
}

// Main

int main(int argc, char **argv)
{
    ros::init(argc, argv, NAME);
    ros::NodeHandle n;

    ROS_INFO("Initializing node '%s' ...", NAME);

    // Read Params
    n.param<double>("rate", rate, RATE);
    secs = 1.0 / rate;
    ROS_INFO("Time between frames: %f seconds.", secs);

    // Initialize Variables
    last_sent = 0;

    // Advertise Publishers
    pub_rgb_info   = n.advertise<sensor_msgs::CameraInfo>(RGB_INFO_OUT,   BUFFER_OUT);
    pub_rgb_rect   = n.advertise<sensor_msgs::Image     >(RGB_RECT_OUT,   BUFFER_OUT);
    pub_depth_info = n.advertise<sensor_msgs::CameraInfo>(DEPTH_INFO_OUT, BUFFER_OUT);
    pub_depth_rect = n.advertise<sensor_msgs::Image     >(DEPTH_RECT_OUT, BUFFER_OUT);

    ros::Subscriber sub_rgb_info   = n.subscribe(RGB_INFO_IN,   BUFFER_IN, callback_rgb_info);
    ros::Subscriber sub_rgb_rect   = n.subscribe(RGB_RECT_IN,   BUFFER_IN, callback_rgb_rect);
    ros::Subscriber sub_depth_info = n.subscribe(DEPTH_INFO_IN, BUFFER_IN, callback_depth_info);
    ros::Subscriber sub_depth_rect = n.subscribe(DEPTH_RECT_IN, BUFFER_IN, callback_depth_rect);

    ros::spin();
    return 0;
}

File: src/rgbd_throttle.py
#!/usr/bin/env python
#*********************************************************************
#*
#*  Copyright (c) 2016, Antonio Coratelli
#*
#*  Software License Agreement (BSD License)
#*
#*  Redistribution and use in source and binary forms, with or without
#*  modification, are permitted provided that the following conditions
#*  are met:
#*
#*   * Redistributions of source code must retain the above copyright
#*     notice, this list of conditions and the following disclaimer.
#*   * Redistributions in binary form must reproduce the above
#*     copyright notice, this list of conditions and the following
#*     disclaimer in the documentation and/or other materials provided
#* 		 the distribution.
#*   * Neither the name of the athor nor the names of its
#*     contributors may be used to endorse or promote products derived
#*     from this software without specific prior written permission.
#*
#*  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
#*  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
#*  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
#*  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
#*  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
#*  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
#*  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
#*  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
#*  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
#*  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
#*  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
#*  POSSIBILITY OF SUCH DAMAGE.
#*
#*  Author: Antonio Coratelli <antoniocoratelli@gmail.com>
#*  Last edit: May, 2016
#*********************************************************************/

PKG  = "rgbd_throttle"
NAME = "rgbd_throttle"

import roslib; roslib.load_manifest(PKG)
import rospy
import tf
import time
from sensor_msgs.msg import CameraInfo
from sensor_msgs.msg import Image

class rgbd_throttle:


    def __init__(self):
        
        # initialize ros node
        rospy.init_node(NAME)
        rospy.loginfo("Initializing node '%s' ...", NAME)
        
        # subscribed topics
        self.rgb_info_in   = 'rgb/info_in'
        self.rgb_rect_in   = 'rgb/rect_in'
        self.depth_info_in = 'depth/info_in'
        self.depth_rect_in = 'depth/rect_in'
        
        # published topics
        self.rgb_info_out   = 'rgb/info_out'
        self.rgb_rect_out   = 'rgb/rect_out'
        self.depth_info_out = 'depth/info_out'
        self.depth_rect_out = 'depth/rect_out'
        
        # get ros params
        self.secs = float(rospy.get_param('rate', 1.0)) ** (-1)
        
        # init variables
        self.last_rgb_info = 0
        self.accept_rgb_info   = True
        self.accept_rgb_rect   = False
        self.accept_depth_info = False
        self.accept_depth_rect = False
        
        # echo info
        rospy.loginfo("Time between frames: %s secs", self.secs)
        
        # declare subscribed topics
        rospy.Subscriber(self.rgb_info_in,   CameraInfo, self.callback_rgb_info)
        rospy.Subscriber(self.rgb_rect_in,   Image,      self.callback_rgb_rect)
        rospy.Subscriber(self.depth_info_in, CameraInfo, self.callback_depth_info)
        rospy.Subscriber(self.depth_rect_in, Image,      self.callback_depth_rect)
        
        self.pub_rgb_info   = rospy.Publisher(self.rgb_info_out,   CameraInfo, queue_size=5)
        self.pub_rgb_rect   = rospy.Publisher(self.rgb_rect_out,   Image,      queue_size=5)
        self.pub_depth_info = rospy.Publisher(self.depth_info_out, CameraInfo, queue_size=5)
        self.pub_depth_rect = rospy.Publisher(self.depth_rect_out, Image,      queue_size=5)
        
        rospy.spin()
    
    
    def __del__(self):
        
        rospy.loginfo("Closing node '%s'", NAME)
    
    
    def callback_rgb_info(self, data):
        
        if (data.header.stamp.to_sec() <  self.last_rgb_info):
            self.last_rgb_info = 0
        
        if (data.header.stamp.to_sec() >= self.last_rgb_info + self.secs):
            self.accept_rgb_info   = True
            self.accept_rgb_rect   = False
            self.accept_depth_info = False
            self.accept_depth_rect = False
        
        if (self.accept_rgb_info):
            # rospy.loginfo("Publishing rgb_info %s", data.header.stamp.to_sec())
            self.pub_rgb_info.publish(data)
            self.last_rgb_info     = data.header.stamp.to_sec()
            self.accept_rgb_info   = False
            self.accept_rgb_rect   = True
            self.accept_depth_info = True
            self.accept_depth_rect = False
    
    def callback_rgb_rect(self, data):
        
        if (self.accept_rgb_rect):
            # rospy.loginfo("Publishing rgb_rect %s", data.header.stamp.to_sec())
            self.pub_rgb_rect.publish(data)
            self.accept_rgb_rect = False
    
    
    def callback_depth_info(self, data):
        
        if (self.accept_depth_info):
            # rospy.loginfo("Publishing depth_info %s", data.header.stamp.to_sec())
            self.pub_depth_info.publish(data)
            self.accept_depth_info = False
            self.accept_depth_rect = True
    
    
    def callback_depth_rect(self, data):
        
        if (self.accept_depth_rect):
            # rospy.loginfo("Publishing depth_rect %s", data.header.stamp.to_sec())
            self.pub_depth_rect.publish(data)
            self.accept_depth_rect = False
    
    
### MAIN #######################################################################

if __name__ == "__main__":

    c = rgbd_throttle()

